Northstar.CustomServers/mod/scripts/vscripts/mp/_stats.nut

```diff
@@ -12,67 +14,1083 @@ global function PreScoreEventUpdateStats
 global function PostScoreEventUpdateStats
 global function Stats_OnPlayerDidDamage
 
+struct {
+	table< string, array<string> > refs
+	table< string, array< void functionref( entity, float, string ) > > callbacks
+
+	table< entity, table< string, int > > cachedIntStatChanges
+	table< table< string, float > > cachedFloatStatChanges
+
+	table< entity, float > playerKills
+	table< entity, float > playerKillsPvp
+	table< entity, float > playerDeaths
+	table< entity, float > playerDeathsPvp
+
+	bool isFirstStrike = true
+} file
+
 void function Stats_Init()
 {
+	AddCallback_OnPlayerKilled( OnPlayerOrNPCKilled )
+	AddCallback_OnNPCKilled( OnPlayerOrNPCKilled )
+	AddCallback_OnPlayerRespawned( OnPlayerRespawned )
+	AddCallback_OnClientConnected( OnClientConnected )
+	AddCallback_OnClientDisconnected( OnClientDisconnected )
+	AddCallback_GameStateEnter( eGameState.Epilogue, OnEpilogueStarted )
+
+	thread HandleDistanceAndTimeStats_Threaded()
+	thread SaveStatsPeriodically_Threaded()
+}
+
+void function AddStatCallback( string statCategory, string statAlias, string statSubAlias, void functionref( entity, float, string ) callback, string subRef )
+{	
+	if ( !IsValidStat( statCategory, statAlias, statSubAlias ) )
+		throw "INVALID STAT: " + statCategory + " : " + statAlias + " : " + statSubAlias
+	
+	
+	string str = GetStatVar( statCategory, statAlias, statSubAlias )
 
+	if ( str in file.refs )
+	{
+		file.refs[str].append( subRef )
+		file.callbacks[str].append( callback )
+	}
+	else
+	{
+		file.refs[str] <- [subRef]
+		file.callbacks[str] <- [callback]
+	}
 }
 
-void function AddStatCallback(string statCategory, string statAlias, string statSubAlias, void functionref(entity, float, string) callback, string subRef)
+// a lot of this file seems to be doing caching of stats in some way
+void function Stats_SaveStatDelayed( entity player, string statCategory, string statAlias, string statSubAlias )
 {
+	// idk how long the delay is meant to be but whatever
+	WaitFrame()
```

Comment:
Otherwise this is kind of pointless tbh
```suggestion
void function Stats_SaveStatDelayed( entity player, string statCategory, string statAlias, string statSubAlias, float delay = 0.1 )
{
	// idk how long the delay is meant to be but whatever
	wait delay
```